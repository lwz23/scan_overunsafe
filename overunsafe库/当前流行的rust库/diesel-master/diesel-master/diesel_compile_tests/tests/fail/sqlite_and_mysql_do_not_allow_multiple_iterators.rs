extern crate diesel;

use diesel::prelude::*;
use diesel::connection::{LoadConnection, DefaultLoadingMode};
use diesel::sql_query;
use diesel::pg::PgRowByRowLoadingMode;

fn main() {
    let conn = &mut SqliteConnection::establish("foo").unwrap();
    // For sqlite the returned iterator is coupled to
    // a statement, which is coupled to the connection itself
    // so we cannot have more than one iterator
    // for the same connection
    let row_iter1 = LoadConnection::load(conn, sql_query("bar")).unwrap();
    let row_iter2 = LoadConnection::load(conn, sql_query("bar")).unwrap();

    let _ = row_iter1.zip(row_iter2);

    let conn = &mut MysqlConnection::establish("foo").unwrap();
    // The same argument applies to mysql
    let row_iter1 = LoadConnection::load(conn, sql_query("bar")).unwrap();
    let row_iter2 = LoadConnection::load(conn, sql_query("bar")).unwrap();

    let _ = row_iter1.zip(row_iter2);

    let conn = &mut PgConnection::establish("foo").unwrap();
    // It works for PgConnection as the result is not related to the
    // connection in any way
    let row_iter1 = LoadConnection::<DefaultLoadingMode>::load(conn, sql_query("bar")).unwrap();
    let row_iter2 = LoadConnection::<DefaultLoadingMode>::load(conn, sql_query("bar")).unwrap();

    let _ = row_iter1.zip(row_iter2);

    // It does not work for the libpq row by row mode
    let row_iter1 = LoadConnection::<PgRowByRowLoadingMode>::load(conn, sql_query("bar")).unwrap();
    let row_iter2 = LoadConnection::<PgRowByRowLoadingMode>::load(conn, sql_query("bar")).unwrap();

    let _ = row_iter1.zip(row_iter2);
}
