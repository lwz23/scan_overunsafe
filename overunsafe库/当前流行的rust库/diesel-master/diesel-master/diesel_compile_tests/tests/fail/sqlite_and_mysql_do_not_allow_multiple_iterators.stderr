error[E0499]: cannot borrow `*conn` as mutable more than once at a time
  --> tests/fail/sqlite_and_mysql_do_not_allow_multiple_iterators.rs:15:42
   |
14 |     let row_iter1 = LoadConnection::load(conn, sql_query("bar")).unwrap();
   |                                          ---- first mutable borrow occurs here
15 |     let row_iter2 = LoadConnection::load(conn, sql_query("bar")).unwrap();
   |                                          ^^^^ second mutable borrow occurs here
16 |
17 |     let _ = row_iter1.zip(row_iter2);
   |             --------- first borrow later used here

error[E0499]: cannot borrow `*conn` as mutable more than once at a time
  --> tests/fail/sqlite_and_mysql_do_not_allow_multiple_iterators.rs:22:42
   |
21 |     let row_iter1 = LoadConnection::load(conn, sql_query("bar")).unwrap();
   |                                          ---- first mutable borrow occurs here
22 |     let row_iter2 = LoadConnection::load(conn, sql_query("bar")).unwrap();
   |                                          ^^^^ second mutable borrow occurs here
23 |
24 |     let _ = row_iter1.zip(row_iter2);
   |             --------- first borrow later used here

error[E0499]: cannot borrow `*conn` as mutable more than once at a time
  --> tests/fail/sqlite_and_mysql_do_not_allow_multiple_iterators.rs:36:67
   |
35 |     let row_iter1 = LoadConnection::<PgRowByRowLoadingMode>::load(conn, sql_query("bar")).unwrap();
   |                                                                   ---- first mutable borrow occurs here
36 |     let row_iter2 = LoadConnection::<PgRowByRowLoadingMode>::load(conn, sql_query("bar")).unwrap();
   |                                                                   ^^^^ second mutable borrow occurs here
37 |
38 |     let _ = row_iter1.zip(row_iter2);
   |             --------- first borrow later used here
